mesa/agent.py

```diff
@@ -57,3 +61,73 @@ def advance(self) -> None:
     @property
     def random(self) -> Random:
         return self.model.random
+
+
+class AgentSet:
+    def __init__(self, agents: Iterable[Agent], model: Model):
+        self._agents = weakref.WeakKeyDictionary()
+
+        for agent in agents:
+            self._agents[agent] = None
+
+        self.model = model
+
+    def __len__(self) -> int:
+        return len(self._agents)
+
+    def __iter__(self) -> Iterator[Agent]:
+        return iter(self._agents.keys())
+
+    def __contains__(self, agent: Agent) -> bool:
+        """Check if an agent is in the AgentSet."""
+        return agent in self._agents
+
+    def select(self, filter_func: Callable[[Agent], bool] | None = None, inplace: bool = False) -> AgentSet:
+        if filter_func is None:
+            if inplace:
+                return self
+            else:
+                return AgentSet(list(self._agents.keys()), self.model)
+        else:
+            agents = [agent for agent in self._agents.keys() if filter_func(agent)]
+
+            if inplace:
+                self._reorder(agents)
+                return self
+            else:
+                return AgentSet(
+                    agents,
+                    self.model
+                )
```

Comment:
I think this code can be improved a bit, it's a bit hard to read at the moment. I'm on mobile so please check this again, but here is what ChatGPT made of it
```python
def select(self, filter_func: Callable[[Agent], bool] | None = None, inplace: bool = False) -> AgentSet:
    """
    Select agents based on a filter function.

    Args:
    filter_func: A callable that returns True for agents to be included.
    inplace: If True, modify the current AgentSet; otherwise, return a new one.

    Returns:
    AgentSet: A new or modified AgentSet based on the filter criteria.
    """

    # Handle the case where no filter function is provided
    if filter_func is None:
        return self if inplace else AgentSet(list(self._agents.keys()), self.model)

    # Apply filter function to select agents
    agents = [agent for agent in self._agents.keys() if filter_func(agent)]

    # Modify in place or return a new AgentSet
    if inplace:
        self._reorder(agents)
        return self
    else:
        return AgentSet(agents, self.model)
```

Reply:
chatgpt version does indeed look a bit nicer.